subsys/caf/modules/sensor_sampler.c

```diff
@@ -61,26 +80,142 @@ static void send_sensor_event(const char *descr, struct sensor_value *data, size
 	EVENT_SUBMIT(event);
 }
 
-static int sample_sensor(const struct device *dev, const struct sensor_config *sc)
+static struct sensor_data *get_sensor_data(const struct device *dev)
+{
+	struct sensor_data *sd = NULL;
+
+	for (size_t i = 0; i < ARRAY_SIZE(sensor_data); i++) {
+		sd = sensor_data + i;
+		if (sd->dev == dev) {
+			break;
+		}
+	}
+
+	__ASSERT_NO_MSG(sd);
+
+	return sd;
+}
+
+static size_t get_sensor_data_cnt(const struct sensor_config *sc)
 {
 	size_t data_cnt = 0;
 
 	for (size_t i = 0; i < sc->chan_cnt; i++) {
 		data_cnt += sc->chans[i].data_cnt;
 	}
 
+	return data_cnt;
+}
+
+static bool is_active(const struct sensor_config *sc, double curr, double prev)
+{
+	enum trigger_activation_type type = sc->trigger->activation.type;
+	double tresh = sc->trigger->activation.tresh;
+
+	switch (type) {
+	case PERCENTAGE_DEVIATION:
+		return FABS(curr - prev) < FABS(prev * tresh/100.0);
+	case ABSOLUTE_DEVIATION:
+		return FABS(curr - prev) < FABS(tresh);
+	default:
+		report_error();
+	}
+
+	return false;
+}
+
+static bool can_sensor_sleep(const struct sensor_config *sc,
+			     struct sensor_data *sd,
+			     struct sensor_value *new)
+{
+	size_t data_cnt = get_sensor_data_cnt(sc);
+	bool sleep = true;
+
+	for (size_t i = 0; i < data_cnt; i++) {
+		double prev = sensor_value_to_double(&sd->prev[i]);
+		double curr = sensor_value_to_double(&new[i]);
+
+		if (!is_active(sc, curr, prev)) {
+			sleep = false;
+			break;
+		}
+	}
+
+	if (sleep) {
+		unsigned int timeout = sc->trigger->activation.timeout_ms;
+		unsigned int period = sc->sampling_period_ms;
+
+		sd->sleep_cnt++;
+		if (sd->sleep_cnt >= timeout/period) {
+			sd->sleep_cnt = 0;
+			return true;
+		}
+	} else {
+		memcpy(sd->prev, new, data_cnt*sizeof(struct sensor_value));
+		sd->sleep_cnt = 0;
+	}
+
+	return false;
+}
+
+static void trigger_handler(const struct device *dev, struct sensor_trigger *const trigger)
```

Comment:
why `struct sensor_trigger *const`?
it should be `const struct  sensor_trigger *`. Making function argument const has little sense.

Reply:
Mistake.